module/executiondatasync/tracker/storage.go

```diff
@@ -0,0 +1,454 @@
+package tracker
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/hashicorp/go-multierror"
+	"github.com/ipfs/go-cid"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/module/blobs"
+)
+
+// badger key prefixes
+const (
+	prefixGlobalState  byte = iota + 1 // global state variables
+	prefixLatestHeight                 // tracks the latest height containing each blob
+	prefixBlobRecord                   // tracks the set of blobs at each height
+)
+
+const (
+	globalStateFulfilledHeight = iota + 1 // latest fulfilled block height
+	globalStatePrunedHeight               // latest pruned block height
+)
+
+func retryOnConflict(db *badger.DB, fn func(txn *badger.Txn) error) error {
+	for {
+		err := db.Update(fn)
+		if errors.Is(err, badger.ErrConflict) {
+			continue
+		}
+		return err
+	}
+}
+
+const globalStateKeyLength = 2
+
+func makeGlobalStateKey(state byte) []byte {
+	globalStateKey := make([]byte, globalStateKeyLength)
+	globalStateKey[0] = prefixGlobalState
+	globalStateKey[1] = state
+	return globalStateKey
+}
+
+const blobRecordKeyLength = 1 + 8 + blobs.CidLength
+
+func makeBlobRecordKey(blockHeight uint64, c cid.Cid) []byte {
+	blobRecordKey := make([]byte, blobRecordKeyLength)
+	blobRecordKey[0] = prefixBlobRecord
+	binary.LittleEndian.PutUint64(blobRecordKey[1:], blockHeight)
+	copy(blobRecordKey[1+8:], c.Bytes())
+	return blobRecordKey
+}
+
+func parseBlobRecordKey(key []byte) (uint64, cid.Cid, error) {
+	blockHeight := binary.LittleEndian.Uint64(key[1:])
+	c, err := cid.Cast(key[1+8:])
+	return blockHeight, c, err
+}
+
+const latestHeightKeyLength = 1 + blobs.CidLength
+
+func makeLatestHeightKey(c cid.Cid) []byte {
+	latestHeightKey := make([]byte, latestHeightKeyLength)
+	latestHeightKey[0] = prefixLatestHeight
+	copy(latestHeightKey[1:], c.Bytes())
+	return latestHeightKey
+}
+
+// getBatchItemCountLimit returns the maximum number of items that can be included in a single batch
+// transaction based on the number / total size of updates per item.
+func getBatchItemCountLimit(db *badger.DB, writeCountPerItem int64, writeSizePerItem int64) int {
+	totalSizePerItem := 2*writeCountPerItem + writeSizePerItem // 2 bytes per entry for user and internal meta
+	maxItemCountByWriteCount := db.MaxBatchCount() / writeCountPerItem
+	maxItemCountByWriteSize := db.MaxBatchSize() / totalSizePerItem
+
+	if maxItemCountByWriteCount < maxItemCountByWriteSize {
+		return int(maxItemCountByWriteCount)
+	} else {
+		return int(maxItemCountByWriteSize)
+	}
+}
+
+type TrackBlobsFn func(uint64, ...cid.Cid) error
```

Comment:
Please add documentation for the exported types and functions. Some suggestions:
* is the height value inclusive or exclusive
* are there any expected errors?